deploy/vagrant/bin/spot_request.py

```diff
@@ -0,0 +1,221 @@
+        chmod 440 /etc/sudoers.d/998-vagrant-cloud-init-requiretty && chmod 440 /etc/sudoers.d/999-vagrant-cloud-init-requiretty"
+
+    requests = conn.request_spot_instances(
+        price = ec2_conf['Spot_Price'],
+        image_id = ec2_conf['AMI'],
+        count = get_init_conf()['MachineNumber'],
+        availability_zone_group = ec2_conf['Availability_Zone'],
+        placement = ec2_conf['Availability_Zone'], # where to put instance
+        key_name = ec2_conf['Keypair'],
+        security_groups = [ec2_conf['Security_Group']],
+        user_data = user_data,
+        instance_type = ec2_conf['Instance_Type'],
+        block_device_map = get_bdm(ec2_conf))
+    request_ids = [r.id for r in requests]
+    save_request_ids(request_ids)
+
+    # block, waiting for all requests to be fulfilled
+    requests = wait_until_fulfilled(request_ids, conn)
+
+    # tag the requests and instances
+    rid_tag = request_id_to_tag(requests)
+    for r in requests:
+        tag = rid_tag[r.id]
+        r.add_tag('Name', tag)
+        conn.create_tags([r.instance_id], {'Name': tag})
+
+    return requests
+
+
+def cancel_request(conn):
+    warn('canceling spot instance requests and terminating instances...')
+    requests = conn.get_all_spot_instance_requests(load_request_ids())
+    for r in requests:
+        r.cancel()
+    instance_ids = [r.instance_id for r in requests if r.instance_id is not None]
+    if len(instance_ids) > 0:
+        conn.terminate_instances(instance_ids)
+
+
+# mock the inventory file and machine id files that should have
+# been generated by vagrant, so that we can keep the vagrant work flow.
+def mock_vagrant_info(instance_id_to_tag_ip):
+    inventory_dir = '.vagrant/provisioners/ansible/inventory'
+    mkdir_p(inventory_dir)
+    inventory = open(os.path.join(inventory_dir, 'vagrant_ansible_inventory'), 'w')
+    for instance_id, tag_ip in instance_id_to_tag_ip.iteritems():
+        tag, ip = tag_ip
+        host = get_host(tag)
+
+        inventory.write("{} ansible_ssh_host={} ansible_ssh_port=22\n".format(host, ip))
+
+        id_dir = os.path.join('.vagrant', 'machines', host, 'aws')
+        mkdir_p(id_dir)
+        with open(os.path.join(id_dir, 'id'), 'w') as f:
+            f.write(instance_id)
+    inventory.close()
+
+
+def is_ssh_ready(host):
+    s = subprocess.Popen(['ssh', 
+        '-o', 'StrictHostKeyChecking=no', 
+        '-o', 'UserKnownHostsFile=/dev/null',
+        '-o', 'ConnectTimeout=30',
+        '-i', os.path.expanduser(get_ec2_conf()['Key_Path']),
+        '%s@%s' % ('ec2-user', host),
+        'true'], 
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT)
+    s.communicate()
+    return s.returncode == 0
+
+
+def wait_for_ssh(hosts):
+    while len(hosts):
+        remove = set()
+        for host in hosts:
+            if is_ssh_ready(host):
+                remove.add(host)
+        hosts -= remove
```

Comment:
the pythonic way `hosts = [h for h in hosts if not is_ssh_ready(h)]`, thanks for hint.